lib/Conversion/CoreToFSM/CoreToFSM.cpp

```diff
@@ -0,0 +1,934 @@
+    product.insert({});
+    return product;
+  }
+
+  // Initialize the product with the elements of the first set, each in its
+  // own vector.
+  for (size_t value : valueSets.front()) {
+    product.insert({value});
+  }
+
+  // Iteratively build the product. For each subsequent set, create a new
+  // temporary product by appending each of its elements to every combination
+  // already generated.
+  for (size_t i = 1; i < valueSets.size(); ++i) {
+    const auto &currentSet = valueSets[i];
+    if (currentSet.empty()) {
+      // The Cartesian product with an empty set results in an empty set.
+      return {};
+    }
+
+    std::set<llvm::SmallVector<size_t>> newProduct;
+    for (const auto &existingVector : product) {
+      for (size_t newValue : currentSet) {
+        llvm::SmallVector<size_t> newVector = existingVector;
+        newVector.push_back(newValue);
+        newProduct.insert(std::move(newVector));
+      }
+    }
+    product = std::move(newProduct);
+  }
+
+  return product;
+}
+
+static FrozenRewritePatternSet loadPatterns(MLIRContext &context) {
+
+  RewritePatternSet patterns(&context);
+  for (auto *dialect : context.getLoadedDialects())
+    dialect->getCanonicalizationPatterns(patterns);
+  ICmpOp::getCanonicalizationPatterns(patterns, &context);
+  AndOp::getCanonicalizationPatterns(patterns, &context);
+  XorOp::getCanonicalizationPatterns(patterns, &context);
+  MuxOp::getCanonicalizationPatterns(patterns, &context);
+  ConcatOp::getCanonicalizationPatterns(patterns, &context);
+  ExtractOp::getCanonicalizationPatterns(patterns, &context);
+  AddOp::getCanonicalizationPatterns(patterns, &context);
+  OrOp::getCanonicalizationPatterns(patterns, &context);
+  MulOp::getCanonicalizationPatterns(patterns, &context);
+  hw::ConstantOp::getCanonicalizationPatterns(patterns, &context);
+  TransitionOp::getCanonicalizationPatterns(patterns, &context);
+  StateOp::getCanonicalizationPatterns(patterns, &context);
+  MachineOp::getCanonicalizationPatterns(patterns, &context);
+  patterns.add(pushIcmp);
+  FrozenRewritePatternSet frozenPatterns(std::move(patterns));
+  return frozenPatterns;
+}
+
+static void getReachableStates(llvm::DenseSet<size_t> &visitableStates,
+                               HWModuleOp moduleOp, size_t currentStateIndex,
+                               llvm::SmallVector<seq::CompRegOp> registers,
+                               OpBuilder opBuilder, bool isInitialState) {
+
+  IRMapping mapping;
+  auto clonedBody =
+      llvm::dyn_cast<HWModuleOp>(opBuilder.clone(*moduleOp, mapping));
+
+  llvm::DenseMap<mlir::Value, int> stateMap =
+      intToRegMap(registers, currentStateIndex);
+  Operation *terminator = clonedBody.getBody().front().getTerminator();
+  auto output = dyn_cast<hw::OutputOp>(terminator);
+  int i = 0;
+  llvm::SmallVector<mlir::Value> values;
+  llvm::SmallVector<mlir::Type> types;
+  llvm::DenseMap<int, mlir::Value> regMap;
```

Comment:
Nit: here (and in other places), there are a lot of `mlir::` prefixes, but there's a `using namespace mlir` directive at the start of the file